kernel/src/log_segment.rs

```diff
@@ -0,0 +1,489 @@
+        let mut metadata_opt = None;
+        let mut protocol_opt = None;
+        for batch in data_batches {
+            let (batch, _) = batch?;
+            if metadata_opt.is_none() {
+                metadata_opt = Metadata::try_new_from_data(batch.as_ref())?;
+            }
+            if protocol_opt.is_none() {
+                protocol_opt = Protocol::try_new_from_data(batch.as_ref())?;
+            }
+            if metadata_opt.is_some() && protocol_opt.is_some() {
+                // we've found both, we can stop
+                break;
+            }
+        }
+        match (metadata_opt, protocol_opt) {
+            (Some(m), Some(p)) => Ok((m, p)),
+            (None, Some(_)) => Err(Error::MissingMetadata),
+            (Some(_), None) => Err(Error::MissingProtocol),
+            (None, None) => Err(Error::MissingMetadataAndProtocol),
+        }
+    }
+
+    // Replay the commit log, projecting rows to only contain Protocol and Metadata action columns.
+    fn replay_for_metadata(
+        &self,
+        engine: &dyn Engine,
+    ) -> DeltaResult<impl Iterator<Item = DeltaResult<(Box<dyn EngineData>, bool)>> + Send> {
+        let schema = get_log_schema().project(&[PROTOCOL_NAME, METADATA_NAME])?;
+        // filter out log files that do not contain metadata or protocol information
+        use crate::Expression as Expr;
+        static META_PREDICATE: LazyLock<Option<ExpressionRef>> = LazyLock::new(|| {
+            Some(Arc::new(Expr::or(
+                Expr::column([METADATA_NAME, "id"]).is_not_null(),
+                Expr::column([PROTOCOL_NAME, "minReaderVersion"]).is_not_null(),
+            )))
+        });
+        // read the same protocol and metadata schema for both commits and checkpoints
+        self.replay(engine, schema.clone(), schema, META_PREDICATE.clone())
+    }
+}
+
+pub(crate) struct LogSegmentBuilder<'a> {
+    fs_client: Arc<dyn FileSystemClient>,
+    log_root: &'a Url,
+    checkpoint: Option<CheckpointMetadata>,
+    start_version: Option<Version>,
+    end_version: Option<Version>,
+    no_checkpoint_files: bool,
+    in_order_commit_files: bool,
+}
+impl<'a> LogSegmentBuilder<'a> {
+    pub(crate) fn new(fs_client: Arc<dyn FileSystemClient>, log_root: &'a Url) -> Self {
+        LogSegmentBuilder {
+            fs_client,
+            log_root,
+            checkpoint: None,
+            start_version: None,
+            end_version: None,
+            no_checkpoint_files: false,
+            in_order_commit_files: false,
+        }
+    }
+
+    pub(crate) fn with_checkpoint(mut self, checkpoint: CheckpointMetadata) -> Self {
+        let _ = self.checkpoint.insert(checkpoint);
+        self
+    }
+
+    pub(crate) fn with_start_version(mut self, version: Version) -> Self {
+        let _ = self.start_version.insert(version);
+        self
+    }
+    pub(crate) fn with_end_version(mut self, version: Version) -> Self {
+        let _ = self.end_version.insert(version);
```

Comment:
Why not just `self.end_version = Some(version);` it's a bit more clear about what you are doing and you don't need the `let _ =` syntax.